### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by renmassey** [rasch001/project3#1]
**Neural Networks in CUDA/PTX**

- Super impressive implementing an MNIST classifier using CUDA essentially by hand where most other people would juse use PyTorch or TensorFlow.
- Therefore, it is really cool how you are able to optimize your architecture on the GPU level through parallelization.
- Very nice and clearly defined steps of each part of your architecture with all of the relevant math to back it up.
- Also very nice describing how exactly all of the data is stored within memory as the model progresses.
- Clever use of an accuracy threshold for your iterative process here.
- Really good idea comparing your implementation with a PyTorch implementation to see how well your model performs to the industry standard/baseline.
- It is interesting to see all your weights and biases listed out after training even though they are not decipherable by human readers.
- It is interesting seeing that you stopped at 60% accuracy for your training threshold. Why not go for 80% or 90%? Especially considering how quickly your system was able to go about the process in its current state.
- It was definitely interesting seeing how much more quickly a finely optimized model converges when given enough attention and care. Definitely useful when scaling up to more advanced problems!

**RL Agent for PTX Code Generation**

- Great motivation with the referenced paper for motivation on this experiment, but it would have been nice to spend a little bit of time defining some of the terminology used here for the more uninitiated (SASS and MAC6 specifically).
- That being said, thank you for including an introductory definition on a dynamic grammar.
- The SASS to PTX conversion function is well-defined and well-executed.
- Well defined objective (reward system) and the means to get there.
- Using a simplified objective parameter appears to be pivotal to the success of this project, considering how computationally intensive of a task it is to train a reinforcement learning network.
- Good concession on the work that needs to be done in order to improve the loss function so that it can result as expected.
- Really nice use of charts to describe the specs of each point of the training process; incredibly well organized!

**Simulation of Attention as an Interacting Particle System**

- Very informative report where the complicated math is clearly and well-explained in terms of the papet and measure theory as a whole.
- Really cool seeing how all of the functions you defined at the start work together to produce a working self-attention mapping sequence. 
- Because of the random initialization of your particles, how reproducible are these results? Or does the utilization of a Gaussian distribution take care of this concern?
- The plot and explanation of the plot do a lot to help visualize and internalize the concepts of this paper.
- Looking forward to seeing what applications there are for these findings! It looks super interesting!

**Price:** A+
